WIP: Integration for reading VOX files with vox-format crate

jgraef opened the pull request:
I marked this PR as work-in-progress, because I haven't written any tests yet.

This PR contains integration for the [`vox-format`](https://docs.rs/vox-format) crate. The integration is behing the feature `vox-format`, available for `building-blocks` and `building_blocks_storage`. VOX files can be loaded as `Array3x1<Color>` or `Array3x1<ColorIndex>`:

```rust
use buildings_blocks::storage::vox_format::from_file;

let voxels: Array3x1<ColorIndex> = from_file("test.vox").unwrap();
```

I'm not very happy that I can't implement the conversion traits in `building_blocks_storage`, because users might want to implement their own `VoxBuffer` or `VoxModelBuffer`, and these conversions would help with it.

I tried creating an example, but I'm not very familiar with bevy. And if I write tests, where would I put a test VOX file?

bonsairobo commented on crates/building_blocks_storage/src/vox_format_conversions.rs:
```diff
@@ -0,0 +1,148 @@
+
+/*
+
+These must live in building_blocks_core`
```
Yea you ought to be able to have these in `building_blocks_core`. Is there a problem with that?

jgraef replied:
No, if that's okay, I will do that. I just wanted to minimize the number of places this touches.

bonsairobo replied:
Works for me. There are already features for nalgebra and glam, so you can follow the pattern there.

jgraef replied:
done

bonsairobo commented on crates/building_blocks_storage/src/vox_format_conversions.rs:
```diff
@@ -0,0 +1,148 @@
+
+impl VoxModelBuffer for Array3x1<ColorIndex> {
+    fn new(size: Size) -> Self {
+        Array3x1::fill_with(extent3i_from_vox(size), |_point| ColorIndex::default())
```
This could be faster if you fill the channel first, like:

```rust
let extent = extent3i_from_vox(size);
let channel = Channel::fill(extent.num_points(), ColorIndex::default());
Array3x1::new(extent, channel)
```

I should probably just have an `Array::fill` function.

jgraef replied:
With the new `VoxChannel` trait I don't want to impose a requirement for `VoxChannel: Clone`. Shouldn't this compile down to the same code anyway?

bonsairobo commented:
Hey thanks for contributing!

> I tried creating an example, but I'm not very familiar with bevy. And if I write tests, where would I put a test VOX file?

Probably use the `sdf_mesh` example as a starting point. And you can put any VOX files in the `examples/assets` directory. Maybe we should just move that out of `examples/`, since it's now for tests as well?

jgraef commented on crates/building_blocks_storage/src/vox_format_conversions.rs:
```diff
@@ -0,0 +1,148 @@
+    read_vox_into(reader, &mut buffer)?;
+
+    // TODO: Return error
+    let model = buffer.model.unwrap_or_else(|| panic!("VOX file does not contain a model with index {}", model_index));
```
I wanted to return an error here, but I can also just return `Result<Option<Array3x1<C>, Error>`. Actually I think it makes more sense and I don't have to write an error type myself or import a crate for it.

jgraef replied:
done

jgraef commented:
The new commit addresses the previous points, except that I now don't want to require `VoxChannel: Default`, and thus `Array::fill_with` is used. But I'm open to suggestions.

The `VoxChannel` trait makes it easier to load `.VOX` files into `Array`s with your own channel types. It also avoids the orphan rules, as it's not possible for a user to implement `VoxModelBuffer for Array3x1<C>`. But now I notice that I impose to use `Array3x1`. I think it would be better to allow loading for any `Array3`, but I'm not very familiar with how channels work. @bonsairobo Maybe you can give me a hint, what I need to change.

I implemented a total of **1** tests :laughing:. The `vox-format` crate comes with tests, so IMHO it's enough to ensure the conversion works.

I also added docs with examples, that run as doc-tests.

bonsairobo commented on Cargo.toml:
```diff
@@ -25,7 +25,7 @@ search = ["building_blocks_search"]
 # Integrations.
 dot_vox = ["building_blocks_storage/dot_vox"]
-vox-format = ["building_blocks_storage/vox-format"]
+vox-format = ["building_blocks_storage/vox-format-1", "building_blocks_core/vox-format"]
```
Do you actually need unique names here?

bonsairobo commented on crates/building_blocks_storage/src/vox_format.rs:
```diff
@@ -0,0 +1,215 @@
+//! Provides implementations for loading [MagicaVoxel `.VOX`][magica_voxel] files with the [`vox-format`] crate:
+//!
+//! 1. [`VoxModelBuffer`] implementions for [`Array3x1`] for all channel types that implement [`VoxChannel`].
+//! 1. [`VoxChannel`] implementations for [`ColorIndex`] and [`Color`].
+//! 1. Utility functions [`from_reader`], [`from_slice`], and [`from_file`] to read a model from a `.VOX`
```
Should be 1, 2, 3?

bonsairobo commented:
> I now don't want to require VoxChannel: Clone, and thus Array::fill_with is used. But I'm open to suggestions.

I just added a `Channel::fill_with` that doesn't require `T: Clone`. You should be able to use that like:

```rust
let channel = Channel::fill_with(extent.num_points(), ColorIndex::default);
Array3x1::new(extent, channel)
```

> I think it would be better to allow loading for any Array3, but I'm not very familiar with how channels work. @bonsairobo Maybe you can give me a hint, what I need to change.

Channels are either a `Channel<T>` or a tuple where all elements are channels, i.e. they implement whatever channel traits you need. E.g. `(Channel<A>, Channel<B>)` implements the same channel traits as `Channel<A>`. So if you want to support loading arrays with arbitrary channels, you would need `Array<Chan> where Chan: GetMut + FillChannels`. Right now the `FillChannels` impl does require `Chan::Data: Clone`, which won't work for you. I could make a new trait like `FillChannelsWith` that would make it possible.

I'd also be happy to accept this PR and make it more generic for you.

jgraef commented:
> I'd also be happy to accept this PR and make it more generic for you.

Perfect. Please go ahead :)
